src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -270,144 +291,160 @@ public void copyItemsetAnswer(FormInstance mainInstance, EvaluationContext evalC
-            // common ground between the two), otherwise we can end up failing to
-            // trigger when the ignored context
-            // exists and the used one doesn't
-
-            // TODO It's fishy to mutate the Triggerable here. We might prefer to return a copy of the original Triggerable
-            return qt.changeContextRefToIntersectWithTriggerable(t);
-
-            // note, if the contextRef is unnecessarily deep, the condition will be
-            // evaluated more times than needed
-            // perhaps detect when 'identical' condition has a shorter contextRef,
-            // and use that one instead?
-
-        } else {
-            qt = QuickTriggerable.of(t);
-            unorderedTriggerables.add(qt);
-
-            Set<TreeReference> triggers = t.getTriggers();
-            for (TreeReference trigger : triggers) {
-                List<QuickTriggerable> triggered = triggerIndex.get(trigger);
-                if (triggered == null) {
-                    triggered = new ArrayList<>();
-                    triggerIndex.put(trigger.clone(), triggered);
-                }
-                if (!triggered.contains(qt)) {
-                    triggered.add(qt);
-                }
-            }
+    /**
+     * Adds the provided triggerable to the DAG.
+     * <p>
+     * This method has side-effects:
+     * <ul>
+     *     <li>If a similar triggerable has been already added, its context gets
+     *     intersected with the provided triggerable to cover both using only
+     *     one entry</li>
+     *     <li>This method builds the index of triggerables per trigger ref
+     *     at {@link #triggerablesPerTrigger}</li>
+     * </ul>
+     */
+    Triggerable addTriggerable(Triggerable triggerable) {
+        QuickTriggerable existingQuickTriggerable = findTriggerable(triggerable);
+        if (existingQuickTriggerable != null) {
+            // We found a quick triggerable wrapping a triggerable matching the provided one.
+            // Since Triggerable.equals() doesn't have into account contexts or targets, the
+            // safest thing to do would be to create a new quick triggerable for the incoming
+            // triggerable, but we could cover both triggerables if we compute the intersection
+            // between their respective context and reuse the one we've already stored.
+            existingQuickTriggerable.intersectContextWith(triggerable);
+            return existingQuickTriggerable.getTriggerable();
+        }
 
-            return t;
+        QuickTriggerable newQuickTriggerable = QuickTriggerable.of(triggerable);
+        allTriggerables.add(newQuickTriggerable);
+
+        // Build the triggerable per trigger index
+        Set<TreeReference> triggers = triggerable.getTriggers();
+        for (TreeReference trigger : triggers) {
+            if (!triggerablesPerTrigger.containsKey(trigger))
+                triggerablesPerTrigger.put(trigger, new HashSet<>());
```

Comment:
The original code used to use a clone of the reference as an index. This uses a shared reference. I can't immediately tell whether that's a safe change or not and want to make sure it's deliberate.

Reply:
I changed that because we're not mutating those keys and each time we call `Triggerable.getTriggers()` we get new objects anyway.

If you think this could be problematic, we can clone them again.

Reply:
But do we know for sure that those references can't be mutated by some other process in JavaRosa after the DAG has been built? I find the fact that those references can be mutated a bit scary in general. The index is accessed when triggerables are triggered so if the keys are mutated that could mean the wrong fields being triggered.

Reply:
@ggalmazor traced through this with me and showed me that indeed `getTriggers` takes us into `XPathConditional.getTriggers` which builds new `TreeReference`s from the actual expressions. So this is safe.